New command note

FrankBlack78 opened the pull request:
Hi. I made a draft version for a new command `note`. It is just the class with its methods and does **not** work with the CLI at the moment. Maybe you can review it before I make the changes for the CLI.

I also chose not to create a data-folder due to the fact, that users may enter sensitive information. Therefore the JSON-file containing the data should be placed in the user's home-directory. Please tell me what you think about that.

Thanks in advance.

FrankBlack78 committed:
Merge pull request #1 from IVIURRAY/master

Merge pull request #3 from FrankBlack78/master

FrankBlack78 committed:
Added a draft version for svc_note.py

Just the class with its methods. Not transformed to work with eve yet.

IVIURRAY commented on eve/service/svc_note.py:
```diff
@@ -0,0 +1,172 @@
+        self.path = str(pathlib.Path.home()) + '/test.json'  # rename for production and make it a hidden file
+
+        # check if JSON-file already exists and if not create it with sample entries
+        if os.path.isfile(self.path):
+            pass
```
Not sure if this will be in the final version but there is almost never a reason to use the `pass` keyword in python unless you're testing something. In this case you're better to do;
```python
if not os.path.isfile(self.path):
    # do my thing
```

FrankBlack78 replied:
I always do this. In my mind every if-statement has to have an else-statement too. It's bad habit. Thanks for mentioning.

IVIURRAY commented on eve/service/svc_note.py:
```diff
@@ -0,0 +1,172 @@
+                    pass
+                if priority is not None:
+                    dic["Priority"] = int(priority)
+                else:
+                    pass
```
No need for all these `pass` keywords - they don't add anything to the program

FrankBlack78 replied:
Same. Just bad habit.

IVIURRAY commented on eve/service/svc_note.py:
```diff
@@ -0,0 +1,172 @@
+import pathlib
+
+
+class Note:
```
I would try to structure your class like the below:

```python
class Note:
    def __init__(self):
        self.notes_filepath = os.path.join(pathlib.Path.home(), 'test.json')
        self.notes_file = json.loads(self.notes_filepath)
        
    def add(payload):
        """A method for adding a new note"""
        new_notes = self.notes_file.append(payload)
        self.save(new_notes)

        
    def delete(note_id):
        """A method to delete a note"""
        new_notes = list(filter(self.notes_file, key=lambda note: node['id'] != note_id))
        self.save(new_notes)
     
     def get(note_id):
          """Get a note by id"""
     
    def edit(note_id, **kwargs):
        """Overwrite a note property buy what you pass into the method"""
        note_to_edit = self.get_note(note_id)
        for key, value in kwargs.items():
            note_to_edit[key] = value

    def save(notes):
        """Save the new notes payload to file"""
       

```

FrankBlack78 replied:
Would you mind if I use `input` for creating new notes or altering existing notes? Maybe it is easier for the user to answer 3 input-questions than to put them to the command line like `eve note add '2021-06-01' 2 'Note Text'

I think of something like this:
`Enter a due date for the note (YYYY-MM-DD) [Default = None]:`
`Enter a priority for the note (integer) [Default = None]:`
`Enter a note text for the note [Default = None]:`

IVIURRAY replied:
Yeah I'm totally fine with that, go with what you think works best!

IVIURRAY commented on eve/service/svc_note.py:
```diff
@@ -0,0 +1,172 @@
+    def output(self):
+        # read JSON-file
+        with open(self.path, encoding='utf-8') as f:
+            self.data = json.load(f)
```
It is never good to create an attribute on an object by using `self.data = 'blah'` outside of `__init__`. It mutates the class and means that you've to call the class in a specific order to know if data is populated or not.

Take a look at [this comment](https://github.com/IVIURRAY/yt-eve/pull/5/files#r622444233) about how to structure this to make sure you always have a `self.notes_file`.

FrankBlack78 replied:
That makes sense. Thank you.

IVIURRAY commented on eve/service/svc_note.py:
```diff
@@ -0,0 +1,172 @@
+
+    def del_note(self, uid):
+        # read JSON-file
+        with open(self.path, encoding='utf-8') as f:
+            self.data = json.load(f)
```
You seem to be doing this a lot, it is probably better to read this once in the instantiation of the class

FrankBlack78 replied:
Yeah. I forgot, that I could instantiate the data in the __init__. That was because I first modeled the output-function where I had to convert integers to strings and so on. But I will restructure.

IVIURRAY commented on eve/service/svc_note.py:
```diff
@@ -0,0 +1,172 @@
+        # delete data for given uid
+        self.data = list(filter(lambda i: i["UID"] != int(uid), self.data))
+
+        # assign new uids to data
```
A good way to get unique ids would be this
```python
import uuid
uuid.uuid4()
```

FrankBlack78 replied:
As far as I can see, this module creates pretty large uids which could be very painful to enter them manually in a CLI environment. I think, the usecase for the class I wrote is storing maybe up to 10 messages at the same time.So if I want to delete a message I just have to type `eve note delete 8` rather than `eve note delete 16fd2706-8baf-433b-82eb-8c7fada847da`. Or do I misunderstand something?

IVIURRAY replied:
Ahh yeah, I didn't realise you were editing based on the id! What would be cool would be to print to the console the options or note titles and then allow the user to select them using the arrow keys...let me get you an example!

IVIURRAY replied:
Like [this](https://stackoverflow.com/questions/45022566/create-python-cli-with-select-interface), so you give the user the options of the note title to edit and then they can select that instead of by id. Seems a bit more intuitive?

FrankBlack78 replied:
Hmm. I think I will put this idea aside for the moment to focus a bit. Because of the small scale of the project I don't think I need "real" uids and stay with my way of numbering. I will implement a version with pure CLI first and come back to this idea of navigating the notes later.

IVIURRAY replied:
Yeah no worries, just trying to give you ideas. You go with what you think is best!

IVIURRAY commented on eve/service/svc_note.py:
```diff
@@ -0,0 +1,172 @@
+
+        print(f'Note with uid {uid} deleted.')
+
+    def alter_note(self, uid, due_date=None, priority=None, note_txt=None):
```
You could use the `**kwargs` syntax here to have an arbitrary amount of attribute you can edit on a note.

IVIURRAY replied:
If you're not familiar with `**kwargs` then try this...

```python
def my_func(**kwags):
    print(kwargs)


my_func(param1=1, param2=2, something_else='abc', another_one=False)
```

FrankBlack78 replied:
That is cool. I implemented that for the method `edit`.

IVIURRAY replied:
Yeah it blew my mind when I first played with it, limitless kwargs. There is also a similar setup for *args.

IVIURRAY commented on eve/service/svc_note.py:
```diff
@@ -0,0 +1,172 @@
+        if os.path.isfile(self.path):
+            pass
+        else:
+            self.touch = '''[
+                {
+                    "UID": 1,
+                    "CreationDate": "2021-04-24",
+                    "AlterationDate": null,
+                    "DueDate": "2021-04-30",
+                    "Priority": null,
+                    "NoteTxt": "This is sample entry 1."
+                },
+                {
+                    "UID": 2,
+                    "CreationDate": "2021-04-24",
+                    "AlterationDate": "2021-04-24",
+                    "DueDate": "2021-04-29",
+                    "Priority": 1,
+                    "NoteTxt": "This is sample entry 2."
+                },
+                {
+                    "UID": 3,
+                    "CreationDate": "2021-04-24",
+                    "AlterationDate": "2021-04-24",
+                    "DueDate": "2021-04-28",
+                    "Priority": 2,
+                    "NoteTxt": "This is sample entry 3."
+                }
+            ]'''
```
It might be easier to store your notes as a single `dict`:

Like this, then you can look up the notes by id rather than having to iterate through the list every time to find the UID you want to edit?
```python
{
  'uuid123': {'creationDate': '', 'DueDate': ''},
  'uuid456': {'creationDate': '', 'DueDate': ''},
  'uuid789': {'creationDate': '', 'DueDate': ''}
}
```

FrankBlack78 replied:
Yeah, might be easier, but I struggle a bit at the moment because I want to order the uids for output by DueDate and Priority.

FrankBlack78 replied:
Ok, stackoverflow solved the problem :-)

IVIURRAY commented:
Hey  @FrankBlack78 - I've given you some feedback above about what you could do to neaten this up! 👍 What you've currently is obviously a prototype but this does look like a useful feature and I don't think you're far off making this work!

FrankBlack78 commented:
Thanks for your comments. I will look into them in the next few days. I am pretty new to this (especially classes). I appreciate your help.

IVIURRAY commented:
@FrankBlack78 no worries. Take your time as its always better to understand things than rush stuff to get it done! - Let me know if you want any more help on things 👍

FrankBlack78 committed:
Added new svc_note.py

Implemented the suggested changes to the class. It is still prototype. If the class itself seems fit for implementation, I will add the user-interface via cmd_note.py.

FrankBlack78 commented:
@IVIURRAY I added a new svc_note.py. It is still a prototype. I will implement the user-interface, if the class seems fit for implementation (that are the unresolved topics above).
I got rid of the `pass` statements and restructured the class itself according to your suggestions (except the get-method which I did not need in my opinion).

Would like to hear, what you think.

For the user-interface I would like to build two variants. One with pure command-line-logic and one with a sort of guided interface for the user.

IVIURRAY commented on eve/service/svc_note.py:
```diff
@@ -0,0 +1,161 @@
+        else:
+            # Delete key value from note_data for given key
+            try:
+                del self.note_data[key]
```
I would recommend using .pop(key) here instead

FrankBlack78 replied:
I don't see the difference between `del` and `pop`, but I changed it for now.

IVIURRAY commented on eve/service/svc_note.py:
```diff
@@ -0,0 +1,161 @@
+                print(f'Key {key} deleted...')
+                # Assign new keys to note_data
+                new_keys = list()
+                for i in range(len(self.note_data)):
```
Have you ever used the enumerate keyword, you can do...which will give you the index you're enumerating and the note. Might make this code a little simpler 

```suggestion
                for i, note in enumerate(self.note_data):
```

FrankBlack78 replied:
Thanks for the hint. I didn't knew `enumerate`. Changed it.

IVIURRAY commented:
Hey this looks a lot better now! Really nice work, I left a few more smaller comments this time.

I presume the cli interface would work well in eve so I'm happy for that to be merged in.

What do you mean by the more guided interface? Would this still be inside eve or a separate GUI?

FrankBlack78 commented:
@IVIURRAY Thank you. Unfortunately I cannot see your new comments. But maybe I am looking on the wrong location?

I want to implement the normal cli-workflow e. g. `eve note edit '2021-05-20' 2 'New entry'`. But I also would like to implement something with input-statements, because I think it could be easier for users to know what the variables are. Something like:

`Enter a due date for the note (YYYY-MM-DD) [Default = None]:`
`Enter a priority for the note (integer) [Default = None]:`
`Enter a note text for the note [Default = None]:`

IVIURRAY commented:
@FrankBlack78 take a look at the inline code comments on `svc_note.py` - about the `.pop()` method and using `enumerate`.

I *think* `click` will only allow you to do `eve note edit '2021-05-20' 2 'New entry'` using `@option`...I think if you want to get inline options like that you'll need to find another package to import and use to get your input....I can't think of any off the top of my head....have you tried just chucking in an `input`?

FrankBlack78 commented:
@IVIURRAY Seems I definitely look in the wrong place, because I can't find any new comments.

I didn't try to implement user-inputs yet. I think I will implement the pure CLI version first and take a look if I can make inputs work after that.

IVIURRAY commented:
Yeah give the cli version a crack, that will still be really cool!

FrankBlack78 committed:
Add cmd_note.py

Added cmd_note.py. It is tested and works fine in my opinion.

FrankBlack78 committed:
Updated svc_note.py

Made some minor changes to svc_note.py (e. g. colorcoding for reached due-dates and highlights for outputs).

FrankBlack78 commented:
I made a possible productive version of eve note. It works as intended. If you want, I update the README.md and requirements.txt and call this finished.

IVIURRAY commented on eve/service/svc_note.py:
```diff
@@ -0,0 +1,161 @@
+            # Edit values for given key
+            edit_counter = 0
+            try:
+                self.note_data[key]['DueDate'] = kwargs['DueDate']
```
how about doing this....then you don't have to do a `try...execpt` on every attribute. If you pass another `kwarg` into the method then it will automatically get added to the note without having to change the code.

```python
if key not in self.note_data:
    self.console.log(f"No note found for key {key}")
    return

for item_key, value in kwargs.items():
    self.note_data[key][item_key] = value
```

FrankBlack78 replied:
Oh this is really beautiful. Changed it. Much more elegant than my solution.

IVIURRAY commented on eve/commands/cmd_note.py:
```diff
@@ -0,0 +1,76 @@
+@click.option('-d', '--due_date', help='Due date (YYYY-MM-DD).', required=False, type=str, default=None)
+@click.option('-p', '--priority', help='Priority.', required=False, type=int, default=None)
+@click.pass_context
+def add(ctx, note_text, due_date, priority):
+    """Add notes."""
+    ctx.obj.note.add(note_text, due_date, priority)
+    ctx.obj.note.output()
```
This is nice! - Very neat and easy to read 👍

FrankBlack78 replied:
Thank you :-)

IVIURRAY commented on eve/commands/cmd_note.py:
```diff
@@ -0,0 +1,76 @@
+def edit(ctx, key, due_date=None, priority=None, note_text=None):
+    """Edit notes."""
+    # If due_date, priority and note_text is given
+    if due_date is not None and priority is not None and note_text is not None:
+        ctx.obj.note.edit(key=str(key), DueDate=due_date, Priority=priority, NoteText=note_text)
+    # If due_date and priority is given
+    elif due_date is not None and priority is not None:
+        ctx.obj.note.edit(key=str(key), DueDate=due_date, Priority=priority)
+    # If due_date and note_text is given
+    elif due_date is not None and note_text is not None:
+        ctx.obj.note.edit(key=str(key), DueDate=due_date, NoteText=note_text)
+    # If due_date is given
+    elif due_date is not None:
+        ctx.obj.note.edit(key=str(key), DueDate=due_date)
+    # If priority and note_text is given
+    elif priority is not None and note_text is not None:
+        ctx.obj.note.edit(key=str(key), Priority=priority, NoteText=note_text)
+    # If priority is given
+    elif priority is not None:
+        ctx.obj.note.edit(key=str(key), Priority=priority)
+    # If note_text is given
+    elif note_text is not None:
+        ctx.obj.note.edit(key=str(key), NoteText=note_text)
+    # If nothing is given
+    else:
+        ctx.obj.note.edit(key=str(key))
+    ctx.obj.note.output()
```
hmm, it seems like you've to extend this method everytime you add an extra field. Could you not default these to something sensible then you know you always have them and can get ride of all thes `x is not None` checks.

```python
@cli.command()
@click.option('-k', '--key', help='Key of the note.', required=True, type=int)
@click.option('-d', '--due_date', help='Due date (YYYY-MM-DD).', required=False, type=str, default=SOME_DEFAULT_DUE_DATE)
@click.option('-p', '--priority', help='Priority.', required=False, type=int, default='Medium')
@click.option('-t', '--note_text', help='Note text.', required=False, type=str, default='')
@click.pass_context
def edit(ctx, key, due_date=None, priority=None, note_text=None):
```

IVIURRAY replied:
Notice I've provided defaults so you don't have to check `None` everytime

FrankBlack78 replied:
You are correct. If the data-structure is extended, you have to extend the logic here too. I struggle with that part. Problem is, that I want the user just to enter the attributes that he wants to alter so he doesn't have to input the whole thing again only because he wants to edit the DueDate for example. Therefore everything except the key is **not** required.

The edit-method is build with **kwargs. But if I now put in default values for that attributes, than these default values will be put into the edit-method as well, and the note will be altered for these attributes which is not desired.

I don't really like my solution due to the fact that it is really clunky but it works and I can't find a way to avoid the if-block.

Maybe it is possible to build a get-method on the class and take the values of the original-note if it is not altered?

IVIURRAY replied:
You could leave the defaults to be `None`

then in your logic to add the data to the note you can do....

```python
def edit(ctx, key, due_date=None, priority=None, note_text=None):
    """Edit notes."""
    ctx.obj.note.edit(key=str(key), DueDate=due_date, Priority=priority, NoteText=note_text)

...

# then when you actually edit the note you can do....
for item_key, value in kwargs.item():
    if value is not None:  # here we're checking that the user actually gave us a value
        notes[key][item_key] = value
```

IVIURRAY commented:
@FrankBlack78 - this looks really good and nearly production worthy. I've added some comments [here](https://github.com/IVIURRAY/yt-eve/pull/5/files) so scroll down and see the comments about each section of code.

Mainly I think we can simplify the code a bit rather than checking for every input, so we can use loops to dynamically add an infinate amount of `kwargs` instead of just the set you define.

FrankBlack78 commented:
@IVIURRAY Ok, something is definitely wrong here, because I don't see **any** of your comments. If I follow the link that you provided, I see the following screen: [https://ibb.co/d7YRPQf](https://ibb.co/d7YRPQf)
On the right side is the code of the two files cmd_note.py and svc_note.py (all green, no comments) and on the left side there is only a long black box with nothing in it.

IVIURRAY commented:
@FrankBlack78 ahh man sorry, github was being really annoying there, my comments were stuck in `pending` state - I think you should be able to see them now 🙈

FrankBlack78 commented:
@IVIURRAY Thank goodness. I thought I was to silly to find the thing ;-)

FrankBlack78 committed:
Updated svc_note.py

FrankBlack78 commented:
I put in some highlighting for outputs and reached DueDates.

IVIURRAY commented:
Nice work and the code is a lot simpler to read now! - I added one more comment about how you can simply that logic about handling the users inputs then I think we are good to merge!

FrankBlack78 commented:
Ok, could do that, but than the user cannot change a value from `something` to `None`. Possibly not the most common use-case, but would be nice to have. What do you think?

Edit: He couldn't do that in my solution too, but as we are speaking about it now...

Ah, forget about it. I don't think we need that. I will change the code as you suggested. It is a solid solution.

I will change the readme.md and the requirements.txt as well. Do I have to do something more because there was a merge of another fork in the meantime?

IVIURRAY commented:
You could create a default constant  then  `EVE_NO_USER_INPUT = "EVE_NO_USER_INPUT"` and then set your defaults to that and compare to that when editing notes? Would that work?

FrankBlack78 commented:
I think it is too much effort to implement that. Some attributes are strings, some are integers and I have to check that too. Or I have to change every attribute to strings. 

I think, changing something to `None` is probably not necessary I think.

I will finish this tomorrow.

IVIURRAY commented:
It should be pretty simple. If it does not equal your EVE_DEFAULT_VALUE then you know the user has set something and types don't matter as if it's different then then the user has set it.

FrankBlack78 commented:
What I mean is, if the user - for example inputs `eve note edit -k 1 -d None` than there is an error raised:

`Error: Invalid value for '-p' / '--priority': None is not a valid integer`

That because I define that `priority` has to be an integer.

IVIURRAY commented:
Ahh I see, maybe priority defaults to -1? Then you check for the const or -1

But I would probably argue that by default the priority should be medium or whatever that value is for your implementation. I'm thinking when I enter an event in my calendar they all default to normal priority and not high

FrankBlack78 commented:
On **creation**, DueDate, Priority and NoteText is None by default and I would like to keep it that way. You can specify it on creation or you can edit it with the edit command.

Only in the rare case, that someone would like to change it to `something` and after that back to `None`, I don't see how I could do that because the the input of None on the commandline always will raise an error.

FrankBlack78 committed:
Final commit of setup.py, requirements.txt and README.md

Edited the setup.py to count up the version number.
Edited the requirements.txt with the rich-module.
Edited the README.md with description of the note command.

FrankBlack78 committed:
Final commit of config.py

New entry for the path to the notes-file.

FrankBlack78 committed:
Final commit svc_note.py

Final version of the svc_note.py

FrankBlack78 committed:
Final commit of cmd_note.py

Final version of cmd_note.py.

FrankBlack78 commented:
Changed all files and made my final commit.

IVIURRAY reviewed (approved):
This looks good to me 👍🎉 it seems that you've some merge conflicts though, I think you're going to need to pull the latest master from my repo and merge it into your fork. 

Google, syncing forked repo back to original.

FrankBlack78 committed:
Delete requirements.txt

FrankBlack78 committed:
Create requirements.txt

FrankBlack78 committed:
Merge pull request #2 from IVIURRAY/master

Sync fork

FrankBlack78 committed:
Update requirements.txt

FrankBlack78 commented:
Should be ready to go now :-)

Thank you very much for your time and effort put into my education. I learn a lot with this project.

IVIURRAY commented:
Merged in!! 🎉 🍾 

Thank you so much for taking the time out to implement a new feature into EVE! - I really do apprechaite it and I'm so glad you found this usefule and learned a lot from the process 👍
